### PR TITLE
ROX-13722: [Prefactor] Do not use scanner protos in node.proto

### PR DESCRIPTION
**Pull request by vikin91**
## Description

This PR is a prefactor for #3757. It contains all changes that are safe
to land as they do not affect other parts of the system than RHCOS node
scanning feature.

The changes here are:
- No longer use scanner protos in the `node.proto`
- Add `rhel_content_sets` field in `storage.NodeInventory_Components` - see [Scanner #1053](https://github.com/stackrox/scanner/pull/1053)
- Use more realistic fake data in the fake node inventory

## Checklist

- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added

### Do not apply

- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed
- CI
- Deployed a test cluster and investigated logs for errors

### CODE REVIEW

**Comment by vikin91** [comment]
Current dependencies on/for this PR:
* master
  * **PR #4392** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/4392" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/4392?utm_source=stack-comment).

File: central/sensor/service/pipeline/nodescansv2/pipeline.go [hunk 1]
```diff
@@ -49,7 +49,7 @@ func (p *pipelineImpl) Run(ctx context.Context, clusterID string, msg *central.M
 	}
 
 	// TODO(ROX-12240, ROX-13053): Do something meaningful with the nodeInventory
-	log.Infof("Central received NodeInventory: %+v", nodeInventory)
+	log.Debugf("Central received NodeInventory: %+v", nodeInventory)
```
Comment (Maddosaurus) [inline central/sensor/service/pipeline/nodescansv2/pipeline.go:52]: I see you already incorporated the feedback regarding loglevels. Great job!

That said, we're keeping the pipeline only until #3757 is merged, as `NodeInventory` is still an independent message here, and we depend on this pipeline to print the inventory, correct?

File: compliance/collection/nodeinventorizer/nodeinventory.go [hunk 2]
```diff
@@ -108,6 +109,26 @@ func convertAndDedupRHELComponents(rc *database.RHELv2Components) []*scannerV1.R
...
 
-func makeComponentKey(component *scannerV1.RHELComponent) string {
+func convertExecutables(exe []*scannerV1.Executable) []*storage.NodeInventory_Components_RHELComponent_Executable {
+	arr := make([]*storage.NodeInventory_Components_RHELComponent_Executable, len(exe))
+	for i, executable := range exe {
+		arr[i] = &storage.NodeInventory_Components_RHELComponent_Executable{
+			Path:             executable.Path,
+			RequiredFeatures: nil,
+		}
+		if executable.GetRequiredFeatures() != nil {
```
Comment (Maddosaurus) [inline compliance/collection/nodeinventorizer/nodeinventory.go:119]: Thanks for switchting to getters here, as discussed in the sync review!

**Review by Maddosaurus — APPROVED** [review]
I spotted some well placed small additions, addressing feedback we discussed in the initial review of #3757.
Well done, thanks for the updates!

File: central/sensor/service/pipeline/nodescansv2/pipeline.go (hunk 1 above)
Comment (vikin91) [reply central/sensor/service/pipeline/nodescansv2/pipeline.go:52]: Yes, no changes to the pipeline here as it affects the decision that we need to do in #3757.

The nodescanv2 pipeline is a dud and does nothing except of printing the debug log message. Moreover, it is guarded by the feature-flag - see [here](https://github.com/stackrox/stackrox/blob/b5394616a49b9999da63534840522c80d9277677/central/sensor/service/pipeline/all/factory.go#L68-L70)

File: proto/storage/node.proto [hunk 3]
```diff
@@ -86,11 +84,47 @@ message NodeScan {
...
+        string                     namespace = 1;
+        repeated RHELComponent     rhel_components = 2;
+        message RHELComponent {
+            int64               id = 1;
+            string              name = 2;
+            string              namespace = 3;
+            string              version = 4;
+            string              arch = 5;
+            string              module = 6;
+            repeated string     cpes = 7;
```
Comment (jvdm) [inline proto/storage/node.proto:105]: With `rhel_content_sets` we don't need `cpes` anymore.

**Review by jvdm — CHANGES_REQUESTED** [review]

File: proto/storage/node.proto (hunk 3 above)
Comment (vikin91) [reply proto/storage/node.proto]: Got it removed in [36de34d](https://github.com/stackrox/stackrox/pull/4392/commits/36de34d14c573894a1020512cdc5aeda73b60d31)

File: compliance/collection/nodeinventorizer/nodeinventory.go [hunk 4]
```diff
@@ -108,6 +108,26 @@ func convertAndDedupRHELComponents(rc *database.RHELv2Components) []*scannerV1.R
 	return maps.Values(convertedComponents)
 }
 
-func makeComponentKey(component *scannerV1.RHELComponent) string {
+func convertExecutables(exe []*scannerV1.Executable) []*storage.NodeInventory_Components_RHELComponent_Executable {
+	arr := make([]*storage.NodeInventory_Components_RHELComponent_Executable, len(exe))
+	for i, executable := range exe {
+		arr[i] = &storage.NodeInventory_Components_RHELComponent_Executable{
+			Path:             executable.Path,
```
Comment (rhybrillou) [inline compliance/collection/nodeinventorizer/nodeinventory.go:115]: `executable.GetPath()`

File: compliance/collection/nodeinventorizer/nodeinventory.go [hunk 5]
```diff
@@ -108,6 +108,26 @@ func convertAndDedupRHELComponents(rc *database.RHELv2Components) []*scannerV1.R
...
+			RequiredFeatures: nil,
+		}
+		if executable.GetRequiredFeatures() != nil {
+			arr[i].RequiredFeatures = make([]*storage.NodeInventory_Components_RHELComponent_Executable_FeatureNameVersion, len(executable.GetRequiredFeatures()))
+		}
+		for i2, fnv := range executable.GetRequiredFeatures() {
```
Comment (rhybrillou) [inline compliance/collection/nodeinventorizer/nodeinventory.go:121]: nit: would it make sense to nest this for loop in the previous block after the array allocation ?

File: proto/storage/node.proto [hunk 6]
```diff
@@ -86,11 +84,46 @@ message NodeScan {
...
-    google.protobuf.Timestamp      scan_time  = 3 [(gogoproto.moretags) = 'search:"Node Scan Time,store"'];
-    scannerV1.Components           components = 4;
-    repeated scannerV1.Note        notes      = 5;
+    string                         node_name = 1;
+    google.protobuf.Timestamp      scan_time = 2;
+
+    // Components represents a subset of the scannerV1.Components proto message containing only fields required for RHCOS node scanning
+    // We are not using scannerV1.Components here for the following reasons:
+    // - to avoid conflicts between v1 and scannerV1 APIs when generating the code in central/graphql/resolvers/generated.go
+    // - to not expose scanner v1 API over stackrox graphql API
```
Comment (rhybrillou) [inline proto/storage/node.proto:93]: Should there be a note to keep `scannerV1.Components` and `storage.NodeInventory_Components` in sync ?

**Review by rhybrillou — APPROVED** [review]
Good work.
Some additional bulletproofing can be added in some places.

File: compliance/collection/nodeinventorizer/nodeinventory.go (hunk 4 above)
Comment (vikin91) [reply compliance/collection/nodeinventorizer/nodeinventory.go]: Good point, added in [fdbd8cd](https://github.com/stackrox/stackrox/pull/4392/commits/fdbd8cd232b3fa7ac139ce100ad7d65fb06e7793)

File: compliance/collection/nodeinventorizer/nodeinventory.go [hunk 7]
```diff
@@ -66,33 +66,33 @@ func protoComponentsFromScanComponents(c *nodes.Components) *scannerV1.Component
...
-func convertAndDedupRHELComponents(rc *database.RHELv2Components) []*scannerV1.RHELComponent {
+func convertAndDedupRHELComponents(rc *database.RHELv2Components) []*storage.NodeInventory_Components_RHELComponent {
 	if rc == nil || rc.Packages == nil {
 		log.Warn("No RHEL packages found in scan result")
 		return nil
 	}
 
-	convertedComponents := make(map[string]*scannerV1.RHELComponent, 0)
+	convertedComponents := make(map[string]*storage.NodeInventory_Components_RHELComponent, 0)
 	for i, rhelc := range rc.Packages {
```
[the remark this answers is not in the document]
Comment (vikin91) [reply compliance/collection/nodeinventorizer/nodeinventory.go:83]: Done in [fdbd8cd](https://github.com/stackrox/stackrox/pull/4392/commits/fdbd8cd232b3fa7ac139ce100ad7d65fb06e7793)

File: compliance/collection/nodeinventorizer/nodeinventory.go (hunk 5 above)
Comment (vikin91) [reply compliance/collection/nodeinventorizer/nodeinventory.go]: dded in [fdbd8cd](https://github.com/stackrox/stackrox/pull/4392/commits/fdbd8cd232b3fa7ac139ce100ad7d65fb06e7793)

**Review by md2119 — APPROVED** [review]

**Review by jvdm — APPROVED** [review]
LGTM.

**Comment by vikin91** [comment]
Thank you for all the reviews. I will merge this as it is as this state is a good starting point for further changes.

After meeting with Mandar yesterday, we might need to abandon #3757, readd `node_id` to `NodeInventory` and use a separate message (not Node) to transfer inventory between Sensor and Central. 
I will make necessary experiments and explain why in the next days. Once we agree on that, I will readd the `node_id` in a **followup** to this PR.

**Pull request merged**